docs/developer/file-structure.rst

```diff
@@ -28,58 +29,179 @@ The PyMedPhys repository is generally structured as follows:
+within these categories is organised into levels. Levelling the source code
+helps to prevent circular code dependencies. See diagram below:
 
 .. code-block:: bash
 
    pymedphys/
+   |
    |-- src/pymedphys
    |   |
    |   |-- dicom/
    |   |   |-- __init__.py
    |   |   |
    |   |   |-- _level1/
    |   |   |   |-- __init__.py
-   |   |   |   |-- a_python_source_code_file.py
-   |   |   |   |-- another_python_source_code_file.py
+   |   |   |   |-- d1a.py
+   |   |   |   |-- d1b.py
    |   |   |
    |   |   |-- _level2/
+   |   |   |   |-- __init__.py
+   |   |   |   |-- d2a.py
+   |   |   |   |-- d2b.py
    |   |   |
    |   |   |-- _level3/
+   |   |   |   |-- __init__.py
+   |   |   |   |-- d3a.py
    |   |   |
    |   |   |-- _level4/
+   |   |   |   |-- __init__.py
+   |   |   |   |-- d4a.py
    |   |
    |   |-- gamma/
    |   |   |-- __init__.py
    |   |   |
    |   |   |-- _level1/
    |   |   |   |-- __init__.py
-   |   |   |   |-- and_yet_even_more_python_code.py
+   |   |   |   |-- g1a.py
    |   |   |
    |   |   |-- _level2/
+   |   |   |   |-- __init__.py
+   |   |   |   |-- g2a.py
+   |   |   |   |-- g2b.py
+   |   |   |   |-- g2c.py
    |   |   |
    |   |   |-- _level3/
+   |   |   |   |-- __init__.py
+   |   |   |   |-- g3a.py
    |   |   |
    |   |   |-- _level4/
+   |   |   |   |-- __init__.py
+   |   |   |   |-- g4a.py
    |   |
    |   |-- ...
+   |   
+   |-- ...
+
+Python files within the source code should have descriptive names indicating
+the functions of the code within them. For example, ``dose.py`` in level 1 of
+``dicom`` is so-named because it contains code that interacts with DICOM RT
+Dose files. However, in order to illustrate how levelling works in PyMedPhys,
+the files in the above diagram have been named according to their level and
+module like so:
+
+``<first-letter-of-module><level number><letter-to-differentiate-files-in-the-same-module-and-level>``
+
+E.g. ``g2a.py`` is the first file in level 2 of the ``gamma`` module in the
+above diagram.
+
+The key to levelling is this: **The code contained in files of a particular
+level should only depend on code in files of lower-numbered levels. Code should
+never depend on code within files of the same level, not of higher-numbered
+levels.**
+
+For example, ``g2a.py`` is in level 2, so code in ``g2a.py`` can depend on code
+in ``g1a.py``, because ``g1a.py`` is in level 1 (a lower-numbered level). In
+contrast code in ``g2a.py`` *cannot* depend on code in ``g2b.py`` (which is in
+the same level), ``g3a.py`` or ``g4a.py`` (which are in higher-numbered
+levels).
+
+This philosophy applies across modules as well. For example, although
+``g2a.py`` is in level 2 of the ``gamma`` module, its code can can depend on
+code in ``d1a.py`` of the ``dicom`` module, because ``d1a.py`` is in level 1.
+Similarly, ``g2a.py`` cannot depend on code in ``d2a.py`` of ``dicom`` level 2,
+nor ``d3a.py`` and ``d4a.py`` of ``dicom`` levels 3 and 4.
```

Comment:
This bit actually isn't true. Once I get a chance this evening I'll reword this, and potentially make the module levelling side of things clearer.

Reply:
No worries.

Just as a quick teaser: does the level number not matter when you're importing from another module? I.e. looser restrictions than what I laid out there.

Reply:
Actually, just a bit different. Levels don't matter across modules, but modules themselves are leveled. Unfortunately `layer-linter` doesn't neatly support the concept of "modules on the same level". I described it within the following feature request on `layer-linter`:

https://github.com/seddonym/layer_linter/issues/88

The creator of layer-linter has actually answered that feature request by implementing independent contracts within his new package:

https://import-linter.readthedocs.io/en/latest/contract_types.html